File: QA/Tower_ODR/Tower_ODR/Tower_ODR/CustomersTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tower_ODR
{
    [TestClass]
    public class CustomersTests
    {

        [TestMethod]
        public void Customers_AddNewCustomer()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Add New Customer");
        }


        [TestMethod]
        public void Customers_FiltersCustomerPage()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify filters are working in Customers Page");
        }

        [TestMethod]
        public void Customers_DeleteNewCustomer()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Delete new customer");
        }

        [TestMethod]
        public void Customers_UpdateNewCustomer()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Update new customer");
        }
        [TestMethod]
        public void Webhook_VerifyWebhookSubscriptionFields()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Webhook Subscription Fields");
        }
        [TestMethod]
        public void Webhook_VerifyAddNewWebhook()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Add New Webhook");
        }
        [TestMethod]
        public void Webhook_VerifyEditWebhook()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Edit Webhook");
        }
        [TestMethod]
        public void Webhook_VerifyDeleteWebhook()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Delete Webhook");
        }
        [TestMethod]
        public void Webhook_VerifyCancelNewWebhook()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Cancel New Webhook");
        }
        [TestMethod]
        public void Webhook_VerifyCancelEditWebhook()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify Cancel Edit Webhook");
        }

        [TestMethod]
        public void OpenAPI_VerifyCreateCredential()
        {
            TowerAutomationTests.TestProtractormethod("CustomersTests", "Verify OpenApi Create Credential");
        }
    }
}

File: QA/EndToEndTesting/TestSuites/CustomersTests.js
describe('Test Execution on Mapping Tables - Customers', function () {

    'use strict';
    var pages = {};
    var testData = require('../resources/testData.json');
    pages.Home = require('../pages/Home.js');
    var homePage = new pages.Home();
    pages.Customers = require('../pages/Customers.js');
    var customerPage = new pages.Customers();

    beforeEach(function () {
        homePage.openSearchPage(testData.search.homeUrl);
        homePage.isPageLoaded();
        expect(browser.getCurrentUrl()).toContain(testData.search.homeUrl);
    })

     it('Verify Add New Customer', function () {
         homePage.openSearchPage(testData.search.homeUrl);
         homePage.isPageLoaded();
         expect(browser.getCurrentUrl()).toContain(testData.search.homeUrl);
         homePage.navigateToMappingTablesPage();
         customerPage.isCustomerPageLoaded();
         if (testData.User.Role === "Admin" || testData.User.Role === "SuperAdmin") {
             customerPage.clickOnAddNewCustomer();
             customerPage.enterAddNewCustomerData();
         }
         else if (testData.User.Role === "User")
             customerPage.isAddNewCustomerAvailable();

     });
      
     it('Verify filters are working in Customers Page', function () {
         homePage.navigateToMappingTablesPage();
         customerPage.isCustomerPageLoaded();
         customerPage.filterByFieldName("CustomerID");
         customerPage.clearFilter("CustomerID");
         customerPage.filterByFieldName("CustomerName");
         customerPage.clearFilter("CustomerName");
         customerPage.filterByFieldName("Category");
         customerPage.clearFilter("Category");
         customerPage.filterByFieldName("Application");
         customerPage.clearFilter("Application");
         customerPage.filterByFieldName("Tenant");
         customerPage.clearFilter("Tenant");
         customerPage.filterByInvalidData("CustomerID");
     })


     it('Verify Update new Customer', function () {
         homePage.navigateToMappingTablesPage();
         customerPage.isCustomerPageLoaded();
         customerPage.openNewRecord();
         customerPage.updateNewRecord();

     })

     it('Verify Delete new customer', function () {
         homePage.navigateToMappingTablesPage();
         customerPage.isCustomerPageLoaded();
         customerPage.openNewRecord();
         customerPage.deleteNewRecord();
     })

    it('Verify Webhook Subscription Fields', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.verifyWebhookSubscriptionFiledsDisplayed();
    })

    it('Verify Add New Webhook', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.addNewWebhook();
    })

    it('Verify Edit Webhook', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.editWebhook();
    })

    it('Verify Delete Webhook', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.deleteWebhook();
    })

    it('Verify Cancel New Webhook', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.cancelNewWebhook();
    })

    it('Verify Cancel Edit Webhook', function () {
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        customerPage.openWebhookPage();
        customerPage.cancelEditWebhook();
    })
    
    it('Verify OpenApi Create Credential', function () {
        homePage.openSearchPage(testData.search.homeUrl);
        homePage.isPageLoaded();
        expect(browser.getCurrentUrl()).toContain(testData.search.homeUrl);
        homePage.navigateToMappingTablesPage();
        customerPage.isCustomerPageLoaded();
        if (testData.User.Role === "Admin" || testData.User.Role === "SuperAdmin") {
            customerPage.clickOnAddNewCustomer();
            customerPage.openAPICustomerAddCredential();
        }
        else if (testData.User.Role === "User")
            customerPage.isAddNewCustomerAvailable();

    });

})

